dropshot/src/router.rs

```diff
@@ -128,12 +136,31 @@ impl PathSegment {
                 "{}",
                 "HTTP URI path segment variable missing trailing \"}\""
             );
+
+            let var = &segment[1..segment.len() - 1];
+
+            let (var, pat) = if let Some(index) = var.find(':') {
+                (&var[..index], Some(&var[index + 1..]))
+            } else {
+                (var, None)
+            };
+
             assert!(
-                segment.len() > 2,
-                "HTTP URI path segment variable name cannot be empty"
+                valid_identifier(var),
+                "HTTP URI path segment variable name must be a valid \
+                 identifier: '{}'",
+                var
             );
 
-            PathSegment::Varname((&segment[1..segment.len() - 1]).to_string())
+            if let Some(pat) = pat {
+                assert!(
+                    pat == ".*",
+                    "Only the pattern '.*' is currently supported"
+                );
+                PathSegment::VarnameRegEx(var.to_string(), pat.to_string())
```

Comment:
Reading through more of this PR, it seems like we're doing a lot of special casing for "exactly the wildcard" - not just that this is the only permitted regex, but also things like [checking that no segments exist after the wildcard](https://github.com/oxidecomputer/dropshot/pull/110/files#diff-1d669c288fab8591bd4f3cbcdfb259d761b603299d9a406dd97761344320990aR370).

Seems kinda like those special cases might be hard to undo later - as an example, is it possible to validate that the intersection of N regexes is empty for all possible input paths, to perform a similar collision detection?

Reply:
Yes, I'm familiar with the regex crate. I don't follow your second comment.

Reply:
ah, let me try to re-phrase.

Basically, when I see structures like this:
https://github.com/oxidecomputer/dropshot/blob/ee7fc26698f1ce680a5dcfd4d0e73280367f6cf5/dropshot/src/router.rs#L94-L101

(specifically the `VariableRest` part)

and checks like this:
https://github.com/oxidecomputer/dropshot/blob/ee7fc26698f1ce680a5dcfd4d0e73280367f6cf5/dropshot/src/router.rs#L297-L305

I interpret that check as: "we'd like to stop someone from registering both `/foo/{bar:.*}` and `/foo/baz`" (as covered by tests below) - but in a more general sense, it seems like we could say "we'd like to prevent callers from registering multiple routes that overlap with one another".

When considering regexes beyond just wildcards, this is where my original question came up - we can maintain checks like this, where we handle only "full wildcard", but making the jump to any other type of regex seems like it'll significantly alter this implementation.

I guess I'm asking what direction the ship is steering towards:

- Do we think pure wildcards will be the only supported regex for the foreseeable future?
- Or do we expect other regexes to be supported? (This led to my question about "finding the intersection between multiple regexes", which seemed like the generalized solution to the "do the routes overlap" question).

Reply:
I think I see. I'm going to change `VarnameRegEx` to `VarnameWildcard` to remove the false generalization.

Reply:
I posted my second comment before I saw your response. I don't anticipate other regexes being supported any time soon.